File: server/src/collectors/xrainRadar.js
/**
 * MLIT XRAIN high-resolution rainfall radar (xrain-radar).
 *
 * XRAIN (国土交通省 XバンドMPレーダ雨量情報) is distributed via the MLIT
 * "国土交通省 防災情報提供センター" portal. There is no documented public
 * JSON/GeoJSON nowcast endpoint — the live product is served as map tiles
 * behind the interactive viewer (www.river.go.jp / x-rain). We probe the
 * official portal for reachability and emit a single honest intel pointer;
 * no fabricated rainfall cells are ever produced.
 *
 * Endpoint UNVERIFIED for bulk JSON — best-effort reachability probe only.
 */

import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';
import { fetchHead } from './_liveHelpers.js';

const SOURCE_ID = 'xrain-radar';
const PORTAL_URL = 'https://www.river.go.jp/x/krademapcommon/xrain/index.html';

export default async function collectXrainRadar() {
  let live = false;
  try { live = await fetchHead(PORTAL_URL); } catch { live = false; }

  return intelEnvelope({
    sourceId: SOURCE_ID,
    items: [{
      uid: intelUid(SOURCE_ID, 'portal'),
      title: 'MLIT XRAIN X-band rainfall radar',
      summary: 'High-resolution (250 m / 1 min) MP-radar rainfall product, tile-served via the MLIT disaster-information portal',
      link: PORTAL_URL,
      language: 'ja',
      published_at: new Date().toISOString(),
      tags: ['weather', 'radar', 'rainfall', 'mlit', live ? 'reachable' : 'unreachable'],
      properties: { operator: '国土交通省', reachable: live },
    }],
    live,
    description: 'MLIT XRAIN X-band MP rainfall radar portal (no public JSON nowcast endpoint)',
  });
}

File: native/collectors/infrastructure/sources/xrain_radar.c
/* collectors/infrastructure/sources/xrain_radar.c
 * Port of server/src/collectors/xrainRadar.js — fetchHead reachability
 * probe → ONE intel item (uid = xrain-radar|portal == intelUid(SOURCE_ID,
 * 'portal')). No public JSON nowcast endpoint; 0 data rows is correct
 * (RULE 8). REFERENCE boj_stats.c (lib/probe.h). */
#include "../../../source.h"
#include "../../../lib/probe.h"
#include "../../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>

#define PORTAL_URL "https://www.river.go.jp/x/krademapcommon/xrain/index.html"

static int run(const source_ctx *ctx, intel_sink *sink) {
  int live = probe_head(ctx->http, PORTAL_URL);
  char now[32]; probe_iso_now(now, sizeof now);

  /* tags: ['weather','radar','rainfall','mlit', live?'reachable':...] */
  cJSON *tags = cJSON_CreateArray();
  cJSON_AddItemToArray(tags, cJSON_CreateString("weather"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("radar"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("rainfall"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("mlit"));
  cJSON_AddItemToArray(tags, cJSON_CreateString(live ? "reachable" : "unreachable"));
  char *tj = cJSON_PrintUnformatted(tags);

  /* properties: { operator, reachable } — EXACT JS key order */
  cJSON *p = cJSON_CreateObject();
  cJSON_AddStringToObject(p, "operator", "国土交通省");
  cJSON_AddBoolToObject(p, "reachable", live);
  char *pj = cJSON_PrintUnformatted(p);

  intel_item it = {0};
  it.remote_key   = "portal";              /* → uid xrain-radar|portal */
  it.title        = "MLIT XRAIN X-band rainfall radar";
  it.summary      = "High-resolution (250 m / 1 min) MP-radar rainfall product, tile-served via the MLIT disaster-information portal";
  it.link         = PORTAL_URL;
  it.lang         = "ja";
  it.published_at = now;
  it.tags_json    = tj;
  it.properties_json = pj;
  int rc = sink->emit(sink, &it);

  free(tj); free(pj);
  cJSON_Delete(tags); cJSON_Delete(p);
  fprintf(stderr, "[xrain-radar] probe reachable=%d\n", live);
  return rc >= 0 ? 0 : -1;
}

static const source_def xrain_radar_def = {
  .id = "xrain-radar", .collector = "infrastructure",
  .name = "XRAIN Rain Radar", .name_ja = "XRAIN 高解像度降雨レーダー",
  .update_interval_sec = 60, .run = run };
REGISTER_SOURCE(xrain_radar_def)
